File: src/main/kotlin/t8/BinaryTree.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package t8;

/**
 *
 * @author kamaj
 */
public class BinaryTree {

    private Node root;
    
    public BinaryTree(String rootValue) {
        root = new Node(rootValue);
    }
    public BinaryTree(){
        root = null;
    }
    /*public BinaryTree(String rootValue, BinaryTree left, BinaryTree right){
        root = new Node(rootValue, left, right);
    } */
    
    public void insert(String aData){
        if(this.root == null){
            root = new Node(aData);
        } else {
            setLocator(aData, this);
        }
    }

    private void setLocator(String aData, BinaryTree parent){

        int comparisonValue = aData.compareTo(parent.root.getData());

        if(comparisonValue < 0){
            if(parent.root.left() == null){
                parent.root.setRight(new BinaryTree(aData));
                return;
            } else {
                parent = parent.root.left();
            }
        } else {
            if(parent.root.right() == null){
                parent.root.setRight(new BinaryTree(aData));
                return;
            } else {
                parent = parent.root.right();
            }
        }
        setLocator(aData, parent);
    }
    
    public BinaryTree find(String aData){
        
        return findLocator(aData, this);
    }

    private BinaryTree findLocator(String aData, BinaryTree parent){
        if(parent != null){
            System.out.println("Iiii");
            int comparisonValue = aData.compareTo(parent.root.getData());

            if(comparisonValue < 0){
                findLocator(aData, parent.root.left());
            } else if(comparisonValue > 0){
                findLocator(aData, parent.root.right());
            }
            return parent;
        } else {
            return null;
        }
    }

    public void preOrder() {
        if (root != null) {
            System.out.println(root.getData()+',');
            if (root.left() != null) // pääseeekö vasemmalle?
                root.left().preOrder();
            if (root.right() != null) // pääseekö oikealle?
                root.right().preOrder();
        }

    }
  
    public void setLeft(BinaryTree tree) {
        root.setLeft(tree);
    }

    public void setRight(BinaryTree tree) {
        root.setRight(tree);
    }
}

File: src/main/kotlin/t18/t8I/BinaryTree.kt
package t18.t8I

class BinaryTree {
    private var root: Node? = null

    fun insert(num: Int) {
        root = setValue(num, root, true)
    }

    private fun setValue(testValue: Int, node: Node?, set: Boolean): Node {
        if(node != null){
            val key: Int = node.key
            when {
                testValue < key -> {
                    if(node.left != null){
                        if(!set) {
                            return setValue(testValue, node.left, false)
                        }
                        setValue(testValue, node.left, true)
                    } else if(set){
                        node.left = Node(testValue)
                    }
                }
                testValue > key -> {
                    if(node.right != null){
                        if(!set) {
                            return setValue(testValue, node.right, false)
                        }
                        setValue(testValue, node.right, true)
                    } else if(set){
                        node.right = Node(testValue)
                    }
                }
            }
        } else {
            return Node(testValue)
        }
        return node
    }
}